# Project 1: SAT & ACT Analysis

The first markdown cell in a notebook is a great place to provide an overview of your entire project. You will likely want to at least state your

## Problem Statement

To recommend to the College Board the actions to take to increase the participation rates of high school students taking SAT and ACT exams.

## Executive Summary

The project reviews the data covering scores and participation rates for high school students taking the ACT and SAT exams. The data is ordered by state level and contains the 2017 and 2018 participation rates, Total, Composite and their sub tests scores. The arrangement allows a state-by-state comparison of the participation rates and scores for each test. 


Many states have made it mandatory to take one of the two tests as a requirement for high school graduation and the data set has shown to be aligned with this observation with majority of the students taking either one of the tests. There is a strong negative correlation found between average test score and participation rates. States with high participation rates are more likely to see lower average scores compared to states with lower participation on the same test. Hence, it's not recommended to compare the average scores among states with large participation differences.


### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [1]:
#Imports:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
# Set the graph style
plt.style.use('ggplot')

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [3]:
# Open the 2017 CSV data files
sat_2017 = pd.read_csv('../data/sat_2017.csv')
act_2017 = pd.read_csv('../data/act_2017.csv')

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

First 10 rows of sat_2017

In [4]:
# View SAT Scores
sat_2017.head()

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            5%                                 593   572   1165
1      Alaska           38%                                 547   533   1080
2     Arizona           30%                                 563   553   1116
3    Arkansas            3%                                 614   594   1208
4  California           53%                                 531   524   1055

First 10 rows of act_2017

In [5]:
# View the ACT scores
act_2017.head()

State Participation  English  Math  Reading  Science Composite
0  National           60%     20.3  20.7     21.4     21.0      21.0
1   Alabama          100%     18.9  18.4     19.7     19.4      19.2
2    Alaska           65%     18.7  19.8     20.4     19.9      19.8
3   Arizona           62%     18.6  19.8     20.1     19.8      19.7
4  Arkansas          100%     18.9  19.0     19.7     19.5      19.4

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

Answer: 

For the sat_2017 dataframe, it provides the attributes: 
* State, identified in the State column 
* Participation, identified as the state-by-state participation rate
* Evidence-based reading and writing, average score of Evidence-based reading and writing
* Math scores, average score of Math scores
* Total scores, average score of Total scores

There are a total of 50 states together with the District of Columbia.

For the act_2017 dataframe, it provides the attributes: 
* State, identified in the State column 
* Average participation rate, identified as the state-by-state participation rate
* English, average score of English scores
* Math, average score of Math scores 
* Reading, average score of Reading scores
* Science, average score of Science scores
* Composite scores, average score of Composite scores

There are a total of 50 states together with the District of Columbia and the national average

#### 4a. Does the data look complete? 

Answer:  When compared to the origina dataset, both the 2017 SAT and ACT tests look complete.

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

## SAT Minimum scores and maximum scores

In [6]:
# sat_2017 minimum and max scores
sat_2017[['Participation','Evidence-Based Reading and Writing',
          'Math','Total']].agg(['min','max'])

Participation  Evidence-Based Reading and Writing  Math  Total
min           10%                                 482    52    950
max           96%                                 644   651   1295

The minimum and maximum scores for ERW is correct.

The minimum score for sat_2017 math is 52, but the range is between 468 ~ 651, based on the pdf file the **correct value should be 524**.

In [7]:
sat_2017['Math'].describe()

count     51.000000
mean     547.627451
std       84.909119
min       52.000000
25%      522.000000
50%      548.000000
75%      599.000000
max      651.000000
Name: Math, dtype: float64

In [8]:
sat_2017.loc[sat_2017['Math'] == sat_2017['Math'].min()]

State Participation  Evidence-Based Reading and Writing  Math  Total
20  Maryland           69%                                 536    52   1060

## ACT Minimum scores and maximum scores

Note that composite scores are strings

In [9]:
act_2017['Composite'].min()

'17.8'

In [10]:
act_2017['Composite'].max()

'25.5'

In [11]:
act_2017[['English', 'Math', 'Reading', 'Science', 'Composite']].agg(['min','max'])

English  Math  Reading  Science Composite
min     16.3  18.0     18.1      2.3      17.8
max     25.5  25.3     26.0     24.9      25.5

**act_2017 science minimum score** is incorrect when compared to the pdf file, the correct value should be **23.2**

In [12]:
act_2017[act_2017['Science'] == act_2017['Science'].min()]

State Participation  English  Math  Reading  Science Composite
21  Maryland           28%     23.3  23.1     24.2      2.3      23.6

In [13]:
act_2017.loc[act_2017['State'] == 'Wyoming']

State Participation  English  Math  Reading  Science Composite
51  Wyoming          100%     19.4  19.8     20.8     20.6     20.2x

Wyoming's Composite score is strange with an x in it's score.

#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

### Fixing error for SAT 2017

**The minimum score for sat_2017** math is **52**, but the range is between 468 ~ 651 in the original dataset, the **correct value should be 524**.

In [14]:
min_math_score = sat_2017[sat_2017['Math'] == sat_2017['Math'].min()]
min_math_score

State Participation  Evidence-Based Reading and Writing  Math  Total
20  Maryland           69%                                 536    52   1060

In [15]:
sat_2017.loc[20,'Math'] = 524

In [16]:
sat_2017.loc[20]

State                                 Maryland
Participation                              69%
Evidence-Based Reading and Writing         536
Math                                       524
Total                                     1060
Name: 20, dtype: object

### Fixing error for ACT 2017

In [17]:
act_2017[act_2017['Science'] == act_2017['Science'].min()]

State Participation  English  Math  Reading  Science Composite
21  Maryland           28%     23.3  23.1     24.2      2.3      23.6

**act_2017 science minimum score** is incorrect when compared to the pdf file, the correct value should be **23.2**

In [18]:
act_2017.loc[21, 'Science'] = 23.2
act_2017.loc[21]

State            Maryland
Participation         28%
English              23.3
Math                 23.1
Reading              24.2
Science              23.2
Composite            23.6
Name: 21, dtype: object

### Fixing error for ACT 2017 Composite score

In [19]:
act_2017[act_2017['Composite'] == '20.2x']

State Participation  English  Math  Reading  Science Composite
51  Wyoming          100%     19.4  19.8     20.8     20.6     20.2x

In [20]:
act_2017['Composite'] =act_2017['Composite'].str.replace("20.2x",'20.2')
act_2017['Composite'] = act_2017['Composite'].astype(float)

In [21]:
act_2017['Composite'].dtype

dtype('float64')

#### 5. What are your data types? 
Display the data types of each feature. 

In [22]:
#code
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [23]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   English        52 non-null     float64
 3   Math           52 non-null     float64
 4   Reading        52 non-null     float64
 5   Science        52 non-null     float64
 6   Composite      52 non-null     float64
dtypes: float64(5), object(2)
memory usage: 3.0+ KB


What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

Answer: 

For sat_2017, participation should be a float type not object

For act_2017, participation should be a float type not object and composite should be a float type not object

#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

## Transforming sat_2017 Participation attribute into float type

In [24]:
#code
sat_2017['Participation'] = sat_2017['Participation'].apply(lambda i:float(i.replace('%','')))
sat_2017['Participation']

0       5.0
1      38.0
2      30.0
3       3.0
4      53.0
5      11.0
6     100.0
7     100.0
8     100.0
9      83.0
10     61.0
11     55.0
12     93.0
13      9.0
14     63.0
15      2.0
16      4.0
17      4.0
18      4.0
19     95.0
20     69.0
21     76.0
22    100.0
23      3.0
24      2.0
25      3.0
26     10.0
27      3.0
28     26.0
29     96.0
30     70.0
31     11.0
32     67.0
33     49.0
34      2.0
35     12.0
36      7.0
37     43.0
38     65.0
39     71.0
40     50.0
41      3.0
42      5.0
43     62.0
44      3.0
45     60.0
46     65.0
47     64.0
48     14.0
49      3.0
50      3.0
Name: Participation, dtype: float64

In [25]:
sat_2017['Participation'].dtype

dtype('float64')

## Transforming act_2017 participation and composite into float type

In [26]:
# act_2017['Participation'].map({'%':''})
act_2017['Participation'] = act_2017['Participation'].apply(lambda s:float(s.replace('%','')))
act_2017['Participation']

0      60.0
1     100.0
2      65.0
3      62.0
4     100.0
5      31.0
6     100.0
7      31.0
8      18.0
9      32.0
10     73.0
11     55.0
12     90.0
13     38.0
14     93.0
15     35.0
16     67.0
17     73.0
18    100.0
19    100.0
20      8.0
21     28.0
22     29.0
23     29.0
24    100.0
25    100.0
26    100.0
27    100.0
28     84.0
29    100.0
30     18.0
31     34.0
32     66.0
33     31.0
34    100.0
35     98.0
36     75.0
37    100.0
38     40.0
39     23.0
40     21.0
41    100.0
42     80.0
43    100.0
44     45.0
45    100.0
46     29.0
47     29.0
48     29.0
49     69.0
50    100.0
51    100.0
Name: Participation, dtype: float64

In [27]:
act_2017['Participation'].dtype

dtype('float64')

- Fix any individual values preventing other columns from being the appropriate type.

- Finish your data modifications by making sure the columns are now typed appropriately.

- Display the data types again to confirm they are correct.

In [28]:
#Code:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   State                               51 non-null     object 
 1   Participation                       51 non-null     float64
 2   Evidence-Based Reading and Writing  51 non-null     int64  
 3   Math                                51 non-null     int64  
 4   Total                               51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


In [29]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     float64
 2   English        52 non-null     float64
 3   Math           52 non-null     float64
 4   Reading        52 non-null     float64
 5   Science        52 non-null     float64
 6   Composite      52 non-null     float64
dtypes: float64(6), object(1)
memory usage: 3.0+ KB


#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

## Changing the column names for Sat_2017

Change them to sat_17_participation, sat_17_evidenced_based_reading_writing, sat_17_math, sat_17_total

First change all to lower case

In [30]:
#code
sat_2017 = sat_2017.rename(str.lower, axis='columns')
sat_2017.head()

state  participation  evidence-based reading and writing  math  total
0     Alabama            5.0                                 593   572   1165
1      Alaska           38.0                                 547   533   1080
2     Arizona           30.0                                 563   553   1116
3    Arkansas            3.0                                 614   594   1208
4  California           53.0                                 531   524   1055

In [31]:
sat_2017.columns

Index(['state', 'participation', 'evidence-based reading and writing', 'math',
       'total'],
      dtype='object')

In [32]:
sat_2017 = sat_2017.rename(columns={k:"sat_17_"+k for k in sat_2017.columns[1:]}) # Run once only, if not it'll duplicate
sat_2017.head()

state  sat_17_participation  \
0     Alabama                   5.0   
1      Alaska                  38.0   
2     Arizona                  30.0   
3    Arkansas                   3.0   
4  California                  53.0   

   sat_17_evidence-based reading and writing  sat_17_math  sat_17_total  
0                                        593          572          1165  
1                                        547          533          1080  
2                                        563          553          1116  
3                                        614          594          1208  
4                                        531          524          1055

In [33]:
sat_2017 = sat_2017.rename(columns={"sat_17_evidence-based reading and writing": "sat_17_erw"})
sat_2017.head()

state  sat_17_participation  sat_17_erw  sat_17_math  sat_17_total
0     Alabama                   5.0         593          572          1165
1      Alaska                  38.0         547          533          1080
2     Arizona                  30.0         563          553          1116
3    Arkansas                   3.0         614          594          1208
4  California                  53.0         531          524          1055

## Changing the columns for act 2017

In [34]:
act_2017.head()

State  Participation  English  Math  Reading  Science  Composite
0  National           60.0     20.3  20.7     21.4     21.0       21.0
1   Alabama          100.0     18.9  18.4     19.7     19.4       19.2
2    Alaska           65.0     18.7  19.8     20.4     19.9       19.8
3   Arizona           62.0     18.6  19.8     20.1     19.8       19.7
4  Arkansas          100.0     18.9  19.0     19.7     19.5       19.4

In [35]:
act_2017 = act_2017.rename(str.lower, axis='columns')
act_2017.head()

state  participation  english  math  reading  science  composite
0  National           60.0     20.3  20.7     21.4     21.0       21.0
1   Alabama          100.0     18.9  18.4     19.7     19.4       19.2
2    Alaska           65.0     18.7  19.8     20.4     19.9       19.8
3   Arizona           62.0     18.6  19.8     20.1     19.8       19.7
4  Arkansas          100.0     18.9  19.0     19.7     19.5       19.4

In [36]:
act_2017 = act_2017.rename(columns={k:"act_17_"+k for k in act_2017.columns[1:]})
act_2017.head()

state  act_17_participation  act_17_english  act_17_math  \
0  National                  60.0            20.3         20.7   
1   Alabama                 100.0            18.9         18.4   
2    Alaska                  65.0            18.7         19.8   
3   Arizona                  62.0            18.6         19.8   
4  Arkansas                 100.0            18.9         19.0   

   act_17_reading  act_17_science  act_17_composite  
0            21.4            21.0              21.0  
1            19.7            19.4              19.2  
2            20.4            19.9              19.8  
3            20.1            19.8              19.7  
4            19.7            19.5              19.4

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**|

In [37]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   state                 51 non-null     object 
 1   sat_17_participation  51 non-null     float64
 2   sat_17_erw            51 non-null     int64  
 3   sat_17_math           51 non-null     int64  
 4   sat_17_total          51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


In [38]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   state                 52 non-null     object 
 1   act_17_participation  52 non-null     float64
 2   act_17_english        52 non-null     float64
 3   act_17_math           52 non-null     float64
 4   act_17_reading        52 non-null     float64
 5   act_17_science        52 non-null     float64
 6   act_17_composite      52 non-null     float64
dtypes: float64(6), object(1)
memory usage: 3.0+ KB


|Feature|Type|Dataset|Description|
|---|---|---|---|
|state|object|SAT|State of the data| 
|sat_17_participation|float|SAT|State participation rate in 2017|
|sat_17_erw|integer|SAT|State average Evidence-Based Reading and Writing score in 2017|
|sat_17_math|integer|SAT|State average Math score in 2017|
|sat_17_total|integer|SAT|State average Total score in 2017|
|state|object|ACT|State of the data|
|act_17_participation|float|ACT|State participation rate in 2017|
|act_17_english|float|ACT|State average English score in 2017|
|act_17_math|float|ACT|State average Math score in 2017|
|act_17_reading|float|ACT|State average Reading score in 2017|
|act_17_science|float|ACT|State average Science score in 2017|
|act_17_composite|float|ACT|State average Composite score in 2017|

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [39]:
# Need to remove index 0 as it's the national aggregate
act_2017.head()

state  act_17_participation  act_17_english  act_17_math  \
0  National                  60.0            20.3         20.7   
1   Alabama                 100.0            18.9         18.4   
2    Alaska                  65.0            18.7         19.8   
3   Arizona                  62.0            18.6         19.8   
4  Arkansas                 100.0            18.9         19.0   

   act_17_reading  act_17_science  act_17_composite  
0            21.4            21.0              21.0  
1            19.7            19.4              19.2  
2            20.4            19.9              19.8  
3            20.1            19.8              19.7  
4            19.7            19.5              19.4

In [40]:
act_2017 = act_2017.drop(0, axis='index')
act_2017.head()

state  act_17_participation  act_17_english  act_17_math  \
1     Alabama                 100.0            18.9         18.4   
2      Alaska                  65.0            18.7         19.8   
3     Arizona                  62.0            18.6         19.8   
4    Arkansas                 100.0            18.9         19.0   
5  California                  31.0            22.5         22.7   

   act_17_reading  act_17_science  act_17_composite  
1            19.7            19.4              19.2  
2            20.4            19.9              19.8  
3            20.1            19.8              19.7  
4            19.7            19.5              19.4  
5            23.1            22.2              22.8

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [41]:
sat_2017.columns

Index(['state', 'sat_17_participation', 'sat_17_erw', 'sat_17_math',
       'sat_17_total'],
      dtype='object')

In [42]:
act_2017.columns

Index(['state', 'act_17_participation', 'act_17_english', 'act_17_math',
       'act_17_reading', 'act_17_science', 'act_17_composite'],
      dtype='object')

In [43]:
#Code:
combined_2017 = sat_2017.merge(act_2017,on='state')
combined_2017.head()

state  sat_17_participation  sat_17_erw  sat_17_math  sat_17_total  \
0     Alabama                   5.0         593          572          1165   
1      Alaska                  38.0         547          533          1080   
2     Arizona                  30.0         563          553          1116   
3    Arkansas                   3.0         614          594          1208   
4  California                  53.0         531          524          1055   

   act_17_participation  act_17_english  act_17_math  act_17_reading  \
0                 100.0            18.9         18.4            19.7   
1                  65.0            18.7         19.8            20.4   
2                  62.0            18.6         19.8            20.1   
3                 100.0            18.9         19.0            19.7   
4                  31.0            22.5         22.7            23.1   

   act_17_science  act_17_composite  
0            19.4              19.2  
1            19.9              19.8  
2            19.8              19.7  
3            19.5              19.4  
4            22.2              22.8

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [44]:
#code
combined_2017.to_csv('../data/combined_2017.csv')

## 2018 Data Import and Cleaning

Links to the 2018 ACT and SAT data are provided in the README. These data live in PDFs, and so you'll get to enjoy practicing some *manual* data collection. Save these data as a CSV in your `data` directory, and import, explore, and clean these data in the same way you did above. **Make sure you comment on your steps so it is clear *why* you are doing each process**.

## Importing 2018 SAT DataFrame

In [45]:
sat_2018 = pd.read_csv('../data/sat_2018.csv')
sat_2018.head()

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            6%                                 595   571   1166
1      Alaska           43%                                 562   544   1106
2     Arizona           29%                                 577   572   1149
3    Arkansas            5%                                 592   576   1169
4  California           60%                                 540   536   1076

In [46]:
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [47]:
sat_2018[['Evidence-Based Reading and Writing', 'Math', 'Total']].agg(['min', 'max'])

Evidence-Based Reading and Writing  Math  Total
min                                 480   480    977
max                                 643   655   1298

### Converting Participation into float type

I'll make *Participation rates* will be in float types for all the Data Frames

In [48]:
sat_2018['Participation'] = sat_2018['Participation'].apply(lambda s:float(s.replace('%',""))) # Run once only
sat_2018['Participation']

0       6.0
1      43.0
2      29.0
3       5.0
4      60.0
5     100.0
6     100.0
7     100.0
8      92.0
9      56.0
10     70.0
11     56.0
12    100.0
13     99.0
14     63.0
15      3.0
16      4.0
17      4.0
18      4.0
19     99.0
20     76.0
21     80.0
22    100.0
23      4.0
24      3.0
25      4.0
26     10.0
27      3.0
28     23.0
29     96.0
30     82.0
31     16.0
32     79.0
33     52.0
34      2.0
35      8.0
36     18.0
37     48.0
38     70.0
39     97.0
40     55.0
41      3.0
42      6.0
43     66.0
44      4.0
45     64.0
46     68.0
47     69.0
48     28.0
49      3.0
50      3.0
Name: Participation, dtype: float64

### Renaming the columns to *sat_2018_column*

In [49]:
# Convert to lower case
sat_2018 = sat_2018.rename(str.lower, axis='columns')
sat_2018.head()

state  participation  evidence-based reading and writing  math  total
0     Alabama            6.0                                 595   571   1166
1      Alaska           43.0                                 562   544   1106
2     Arizona           29.0                                 577   572   1149
3    Arkansas            5.0                                 592   576   1169
4  California           60.0                                 540   536   1076

In [50]:
# Adding "sat_18_" to the column names
sat_2018 = sat_2018.rename(columns={col:"sat_18_"+col for col in sat_2018.columns[1:]})
sat_2018.head()

state  sat_18_participation  \
0     Alabama                   6.0   
1      Alaska                  43.0   
2     Arizona                  29.0   
3    Arkansas                   5.0   
4  California                  60.0   

   sat_18_evidence-based reading and writing  sat_18_math  sat_18_total  
0                                        595          571          1166  
1                                        562          544          1106  
2                                        577          572          1149  
3                                        592          576          1169  
4                                        540          536          1076

In [51]:
# Rename evidence-based reading and writing to erw
sat_2018 = sat_2018.rename(columns={"sat_18_evidence-based reading and writing": "sat_18_erw"})
sat_2018.head()

state  sat_18_participation  sat_18_erw  sat_18_math  sat_18_total
0     Alabama                   6.0         595          571          1166
1      Alaska                  43.0         562          544          1106
2     Arizona                  29.0         577          572          1149
3    Arkansas                   5.0         592          576          1169
4  California                  60.0         540          536          1076

In [52]:
sat_2018.tail()

state  sat_18_participation  sat_18_erw  sat_18_math  sat_18_total
46       Virginia                  68.0         567          550          1117
47     Washington                  69.0         543          538          1081
48  West Virginia                  28.0         513          486           999
49      Wisconsin                   3.0         641          653          1294
50        Wyoming                   3.0         633          625          1257

## Importing 2018 ACT DataFrame

In [53]:
act_2018 = pd.read_csv('../data/act_2018_updated.csv')
act_2018.head()

State  Percentage of Students Tested  Average Composite Score  \
0     Alabama                            100                     19.1   
1      Alaska                             33                     20.8   
2     Arizona                             66                     19.2   
3    Arkansas                            100                     19.4   
4  California                             27                     22.7   

   Average English Score  Average Math Score  Average Reading Score  \
0                   18.9                18.3                   19.6   
1                   19.8                20.6                   21.6   
2                   18.2                19.4                   19.5   
3                   19.1                18.9                   19.7   
4                   22.5                22.5                   23.0   

   Average Science Score  
0                   19.0  
1                   20.7  
2                   19.2  
3                   19.4  
4                   22.1

In [54]:
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 7 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   State                          51 non-null     object 
 1   Percentage of Students Tested  51 non-null     int64  
 2   Average Composite Score        51 non-null     float64
 3   Average English Score          51 non-null     float64
 4   Average Math Score             51 non-null     float64
 5   Average Reading Score          51 non-null     float64
 6   Average Science Score          51 non-null     float64
dtypes: float64(5), int64(1), object(1)
memory usage: 2.9+ KB


In [55]:
act_2018.agg(['min', 'max'])

State  Percentage of Students Tested  Average Composite Score  \
min  Alabama                              7                     17.7   
max  Wyoming                            100                     25.6   

     Average English Score  Average Math Score  Average Reading Score  \
min                   16.6                17.8                   18.0   
max                   26.0                25.2                   26.1   

     Average Science Score  
min                   17.9  
max                   24.9

### Transforming act_2018 'Percentage of Students Tested' into float type

Participation rates and percentage of students tested will be converted to floating types

In [56]:
act_2018['Percentage of Students Tested'] = act_2018['Percentage of Students Tested'].astype('float')
act_2018['Percentage of Students Tested'].dtype

dtype('float64')

### Transforming the columns for act_2018

In [57]:
# act_2018 = act_2018.rename(str.lower, axis='columns')
# act_2018.head()

In [58]:
act_2018.columns

Index(['State', 'Percentage of Students Tested', 'Average Composite Score',
       'Average English Score', 'Average Math Score', 'Average Reading Score',
       'Average Science Score'],
      dtype='object')

In [59]:
def add_underscore_char(words,test_name='test'):
    '''
    This function will covert the strings to lower case and add the test and year
    names to the column
    '''
    new_dict = dict()
    for word in words:
        lst = [w.lower() for w in word.split()]
        if len(lst) > 1:
            if lst[0] == 'percentage':
                new_dict[word] = test_name + "participation"
            elif lst[0] == 'average':
                new_dict[word] = test_name + lst[1]
            else:
                new_dict[word] = test_name+"_".join(lst)
        else:
            new_dict[word] = word.lower()
    return new_dict
            
add_underscore_char(act_2018.columns,"act_18_")

{'State': 'state',
 'Percentage of Students Tested': 'act_18_participation',
 'Average Composite Score': 'act_18_composite',
 'Average English Score': 'act_18_english',
 'Average Math Score': 'act_18_math',
 'Average Reading Score': 'act_18_reading',
 'Average Science Score': 'act_18_science'}

In [60]:
act_2018 = act_2018.rename(columns= add_underscore_char(act_2018.columns, "act_18_")) # Run once only

In [61]:
act_2018.head(3)

state  act_18_participation  act_18_composite  act_18_english  \
0  Alabama                 100.0              19.1            18.9   
1   Alaska                  33.0              20.8            19.8   
2  Arizona                  66.0              19.2            18.2   

   act_18_math  act_18_reading  act_18_science  
0         18.3            19.6            19.0  
1         20.6            21.6            20.7  
2         19.4            19.5            19.2

In [62]:
act_2018.tail()

state  act_18_participation  act_18_composite  act_18_english  \
46       Virginia                  24.0              23.9            23.8   
47     Washington                  24.0              22.2            21.4   
48  West Virginia                  65.0              20.3            19.8   
49      Wisconsin                 100.0              20.5            19.8   
50        Wyoming                 100.0              20.0            19.0   

    act_18_math  act_18_reading  act_18_science  
46         23.3            24.7            23.5  
47         22.2            22.7            22.0  
48         19.4            21.3            20.4  
49         20.3            20.6            20.8  
50         19.7            20.6            20.3

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

In [63]:
combined_2018 = sat_2018.merge(act_2018,on='state')
combined_2018.head(3)

state  sat_18_participation  sat_18_erw  sat_18_math  sat_18_total  \
0  Alabama                   6.0         595          571          1166   
1   Alaska                  43.0         562          544          1106   
2  Arizona                  29.0         577          572          1149   

   act_18_participation  act_18_composite  act_18_english  act_18_math  \
0                 100.0              19.1            18.9         18.3   
1                  33.0              20.8            19.8         20.6   
2                  66.0              19.2            18.2         19.4   

   act_18_reading  act_18_science  
0            19.6            19.0  
1            21.6            20.7  
2            19.5            19.2

In [64]:
combined_2017.head(3)

state  sat_17_participation  sat_17_erw  sat_17_math  sat_17_total  \
0  Alabama                   5.0         593          572          1165   
1   Alaska                  38.0         547          533          1080   
2  Arizona                  30.0         563          553          1116   

   act_17_participation  act_17_english  act_17_math  act_17_reading  \
0                 100.0            18.9         18.4            19.7   
1                  65.0            18.7         19.8            20.4   
2                  62.0            18.6         19.8            20.1   

   act_17_science  act_17_composite  
0            19.4              19.2  
1            19.9              19.8  
2            19.8              19.7

In [65]:
final_sat_act = combined_2017.merge(combined_2018, on='state')
final_sat_act.head(3)

state  sat_17_participation  sat_17_erw  sat_17_math  sat_17_total  \
0  Alabama                   5.0         593          572          1165   
1   Alaska                  38.0         547          533          1080   
2  Arizona                  30.0         563          553          1116   

   act_17_participation  act_17_english  act_17_math  act_17_reading  \
0                 100.0            18.9         18.4            19.7   
1                  65.0            18.7         19.8            20.4   
2                  62.0            18.6         19.8            20.1   

   act_17_science  ...  sat_18_participation  sat_18_erw  sat_18_math  \
0            19.4  ...                   6.0         595          571   
1            19.9  ...                  43.0         562          544   
2            19.8  ...                  29.0         577          572   

   sat_18_total  act_18_participation  act_18_composite  act_18_english  \
0          1166                 100.0              19.1            18.9   
1          1106                  33.0              20.8            19.8   
2          1149                  66.0              19.2            18.2   

   act_18_math  act_18_reading  act_18_science  
0         18.3            19.6            19.0  
1         20.6            21.6            20.7  
2         19.4            19.5            19.2  

[3 rows x 21 columns]

In [66]:
final_sat_act.to_csv('../data/final.csv')

Update the data dictionary

|Feature|Type|Dataset|Description|
|---|---|---|---|
|state|object|final.csv|State of the data| 
|sat_17_participation|float|final.csv|State participation rate in 2017|
|sat_17_erw|integer|final.csv|State average Evidence-Based Reading and Writing score in 2017|
|sat_17_math|integer|final.csv|State average Math score in 2017|
|sat_17_total|integer|final.csv|State average Total score in 2017|
|act_17_participation|float|final.csv|State participation rate in 2017|
|act_17_english|float|final.csv|State average English score in 2017|
|act_17_math|float|final.csv|State average Math score in 2017|
|act_17_reading|float|final.csv|State average Reading score in 2017|
|act_17_science|float|final.csv|State average Science score in 2017|
|act_17_composite|float|final.csv|State average Composite score in 2017|
|sat_18_participation|float|final.csv|State participation rate in 2018|
|sat_18_erw|integer|final.csv|State average Evidence-Based Reading and Writing score in 2018|
|sat_18_math|integer|final.csv|State average Math score in 2018|
|sat_18_total|integer|final.csv|State average Total score in 2018|
|act_18_participation|float|final.csv|State participation rate in 2018|
|act_18_english|float|final.csv|State average English score in 2018|
|act_18_math|float|final.csv|State average Math score in 2018|
|act_18_reading|float|final.csv|State average Reading score in 2018|
|act_18_science|float|final.csv|State average Science score in 2018|
|act_18_composite|float|final.csv|State average Composite score in 2018|

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [67]:
final_sat_act.describe().T

count         mean        std    min      25%     50%  \
sat_17_participation   51.0    39.803922  35.276632    2.0     4.00    38.0   
sat_17_erw             51.0   569.117647  45.666901  482.0   533.50   559.0   
sat_17_math            51.0   556.882353  47.121395  468.0   523.50   548.0   
sat_17_total           51.0  1126.098039  92.494812  950.0  1055.50  1107.0   
act_17_participation   51.0    65.254902  32.140842    8.0    31.00    69.0   
act_17_english         51.0    20.931373   2.353677   16.3    19.00    20.7   
act_17_math            51.0    21.182353   1.981989   18.0    19.40    20.9   
act_17_reading         51.0    22.013725   2.067271   18.1    20.45    21.8   
act_17_science         51.0    21.450980   1.739353   18.2    19.95    21.3   
act_17_composite       51.0    21.519608   2.020695   17.8    19.80    21.4   
sat_18_participation   51.0    45.745098  37.314256    2.0     4.50    52.0   
sat_18_erw             51.0   563.686275  47.502627  480.0   534.50   552.0   
sat_18_math            51.0   556.235294  47.772623  480.0   522.50   544.0   
sat_18_total           51.0  1120.019608  94.155083  977.0  1057.50  1098.0   
act_18_participation   51.0    61.647059  34.080976    7.0    28.50    66.0   
act_18_composite       51.0    21.486275   2.106278   17.7    19.95    21.3   
act_18_english         51.0    20.988235   2.446356   16.6    19.10    20.2   
act_18_math            51.0    21.125490   2.035765   17.8    19.40    20.7   
act_18_reading         51.0    22.015686   2.167245   18.0    20.45    21.6   
act_18_science         51.0    21.345098   1.870114   17.9    19.85    21.1   

                          75%     max  
sat_17_participation    66.00   100.0  
sat_17_erw             613.00   644.0  
sat_17_math            599.00   651.0  
sat_17_total          1212.00  1295.0  
act_17_participation   100.00   100.0  
act_17_english          23.30    25.5  
act_17_math             23.10    25.3  
act_17_reading          24.15    26.0  
act_17_science          23.20    24.9  
act_17_composite        23.60    25.5  
sat_18_participation    77.50   100.0  
sat_18_erw             610.50   643.0  
sat_18_math            593.50   655.0  
sat_18_total          1204.00  1298.0  
act_18_participation   100.00   100.0  
act_18_composite        23.55    25.6  
act_18_english          23.70    26.0  
act_18_math             23.15    25.2  
act_18_reading          24.10    26.1  
act_18_science          23.05    24.9

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [115]:
# Standard deviation function
def std(x):
    mean = np.mean(x) # mean of x
    errors = np.square(x - mean) # squared difference
    var = np.sum(errors)/(len(x)-1) # Sum of squared squared difference
    return np.sqrt(var) # Standard deviation

In [116]:
a = final_sat_act['sat_17_participation']
std(a)

35.276632270013046

In [117]:
np.std(a)

34.929070766645076

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [118]:
sd = {attr: round(std(value),2) for attr, value in final_sat_act.select_dtypes(include=['float', 'int']).items()}
sd

{'sat_17_participation': 35.28,
 'act_17_participation': 32.14,
 'act_17_english': 2.35,
 'act_17_math': 1.98,
 'act_17_reading': 2.07,
 'act_17_science': 1.74,
 'act_17_composite': 2.02,
 'sat_18_participation': 37.31,
 'act_18_participation': 34.08,
 'act_18_composite': 2.11,
 'act_18_english': 2.45,
 'act_18_math': 2.04,
 'act_18_reading': 2.17,
 'act_18_science': 1.87}

In [120]:
# round(final_sat_act.select_dtypes(include=['float', 'int']).agg('std'),2)
final_sat_act.std()

sat_17_participation    35.276632
sat_17_erw              45.666901
sat_17_math             47.121395
sat_17_total            92.494812
act_17_participation    32.140842
act_17_english           2.353677
act_17_math              1.981989
act_17_reading           2.067271
act_17_science           1.739353
act_17_composite         2.020695
sat_18_participation    37.314256
sat_18_erw              47.502627
sat_18_math             47.772623
sat_18_total            94.155083
act_18_participation    34.080976
act_18_composite         2.106278
act_18_english           2.446356
act_18_math              2.035765
act_18_reading           2.167245
act_18_science           1.870114
dtype: float64

In [125]:
np.std(final_sat_act.select_dtypes(include=['float', 'int']),ddof=1)

sat_17_participation    35.276632
act_17_participation    32.140842
act_17_english           2.353677
act_17_math              1.981989
act_17_reading           2.067271
act_17_science           1.739353
act_17_composite         2.020695
sat_18_participation    37.314256
act_18_participation    34.080976
act_18_composite         2.106278
act_18_english           2.446356
act_18_math              2.035765
act_18_reading           2.167245
act_18_science           1.870114
dtype: float64

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

Answer: The manually calcuated standard deviations are both the same from the Pandas describe method and the numpy's `std` method.

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

### 2017 SAT Participation rates

In [ ]:
sat_17_part = final_sat_act[['state','sat_17_participation']]
sat_17_part = sat_17_part.sort_values('sat_17_participation', ascending=False) # Sort the participation values
# sat_17_sort_state = sat_17_part.sort_index() # Sort the states

In [ ]:
sat_17_part.head() # Highest participation rates for sat 2017

In [ ]:
sat_17_part.tail() # Lowest participation rates for sat 2017

### SAT 18 Participation rates

In [ ]:
#code
sat_18_part = final_sat_act[['state', 'sat_18_participation']]
sat_18_part = sat_18_part.sort_values('sat_18_participation', ascending=False) # Sort the participation values
sat_18_part.head(10)

In [ ]:
sat_18_part.head() # Highest participation rates for sat 2018

In [ ]:
sat_18_part.tail() # Lowest participation rates for sat 2018

sat 17 states with the highest participation rates:

* District of Columbia
* Michigan
* Connecticut
* Delaware 
* New Hampshire

sat 17 states with lowest participation rates:
* Arkansas
* Wyoming
* North Dakota
* Mississippi
* Iowa

sat 18 states with highest participation rates:

* Colorado
* Connecticut
* Delaware 
* Michigan
* Idaho

sat 18 states with lowest participation rates:
* Mississippi
* Iowa
* Wisconsin
* Wyoming
* North Dakota 	

Participation rates in **Colorado** increased significantly in 2018 compared to 2017. 

### 2017 ACT Participation rates

In [ ]:
act_17_part = final_sat_act[['state','act_17_participation']]
# .set_index(states).drop('state',axis=1)
act_17_part = act_17_part.sort_values('act_17_participation', ascending=False) # Sort the participation values
act_17_part.head() # States with highest participation for act 2017

In [ ]:
act_17_part.tail() # States with lowest participation act 2017

### 2018 ACT Participation rates

In [ ]:
act_18_part = final_sat_act[['state','act_18_participation']]
act_18_part = act_18_part.sort_values('act_18_participation', ascending=False) # Sort the participation values
act_18_part.head() # States with highest participation for act 2018

In [ ]:
act_18_part.tail() # States with lowest participation for act 2018

## Top and bottom 5 SAT 17 and SAT 18 total scores for the states

In [ ]:
sat_17_18_scores = final_sat_act[['state','sat_17_total', 'sat_18_total']].copy()
sat_17_18_scores.head()

In [ ]:
sat_17_scores = sat_17_18_scores.sort_values('sat_17_total', ascending=False).drop('sat_18_total', axis=1)
sat_17_scores.head() # Top 5 SAT total scores for each state 2017

In [ ]:
sat_17_scores.tail() # Bottom 5 SAT total scores for each state 2017

In [ ]:
sat_18_scores = sat_17_18_scores.sort_values('sat_18_total', ascending=False).drop('sat_17_total', axis=1)
sat_18_scores.head() # Top 5 SAT total scores for each state 2018

In [ ]:
sat_18_scores.tail() # Bottom 5 SAT total scores for each state 2018

## Top and bottom 5 ACT 17 and ACT 18 total scores for the states

In [ ]:
final_sat_act.columns

In [ ]:
act_17_18_scores = final_sat_act[['state','act_17_composite', 'act_18_composite']].copy()
act_17_18_scores.head()

In [ ]:
act_17_scores = act_17_18_scores[['state', 'act_17_composite']].sort_values('act_17_composite', ascending=False).copy()
act_17_scores.head() # Highest act 2017 scores

In [ ]:
act_17_scores.tail() # Lowest act 2017 scores

### Do any states with 100% participation on a given test have a rate change year-to-year?

### SAT 2017 and 2018 participation rates

In [ ]:
sat_2017.head()

In [ ]:
sat_2018.head()

In [ ]:
sat_all = sat_2017.merge(sat_2018, on='state')
sat_all.head()

In [ ]:
sat_full_participation = sat_all[(sat_all['sat_17_participation'] == 100) | (sat_all['sat_18_participation'] == 100)].copy()
sat_full_participation.sort_values(['sat_17_participation', 'sat_17_participation'])

The participation rate for **sat** increases significantly for *Colorado* from 11% in 2017 to 100%. This also lead to a decrease in test scores for SAT. 

The other states do not differ as much.

### ACT 2017 and 2018 participation rates

In [ ]:
act_all = act_2017.merge(act_2018, on='state')
act_all.head()

In [ ]:
act_part_17_18 = act_all[(act_all['act_17_participation'] == 100) | (act_all['act_18_participation'] == 100)][['state', 'act_17_participation', 'act_18_participation']]
act_part_17_18.sort_values(['act_17_participation', 'act_18_participation'])

*Colorado* faced the biggest decline in participation rates of ACT 100% in 2017 to **30%** in 2018.

Largest increase in participation rate for *Ohio* with about 33% increase in participation rates

### Do any states show have >50% participation on both tests either year?

In [ ]:
final_sat_act.columns

In [ ]:
final_sat_act[(final_sat_act['sat_17_participation'] > 50) & (final_sat_act['act_17_participation'] > 50)][['state','sat_17_participation','sat_18_participation',
                                                                                                            'act_17_participation', 'act_18_participation']]

Only 3 states, Florida, Georgia and Hawaii showed more than 50% participation rates in both tests and years.

This shows that most students will take either 1 of the tests due to either time or financial constraints.

## In Summary

**SAT tests**

The participation rate for SAT increases significantly for **Colorado** from 11% in 2017 to 100%. This also lead to a decrease in test scores for SAT.

The other states do not differ as much.

**Illinois** has a large increase in SAT participation with a 1000% increase in sat participation from 2017 to 2018


**ACT tests**

**Colorado** faced the biggest decline in participation rates of act 100% in 2017 to 30% in 2018.

Largest increase in participation rate for **Ohio** with about 33% increase in participation rates


**Do any states show have >50% participation on both tests either year?**

Only 3 states, Florida, Georgia and Hawaii showed more than 50% participation rates in both tests and years.

This shows that most students will take either 1 of the tests due to either time or financial constraints.

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

In [ ]:
mask = np.zeros_like(final_sat_act.corr())
mask[np.triu_indices_from(mask)] = True

plt.figure(figsize=(20,9))
sns.heatmap(final_sat_act.corr(),cmap='viridis',annot=True, mask = mask, annot_kws={'size':12})
plt.title('Correlation between ACT and SAT attributes', fontsize=25)
plt.xticks(fontsize=12)
plt.yticks(fontsize=12)

plt.show()

From the heatmap, there's high negative correlation between SAT and ACT participation rate with -84% in 2017 and -85% in 2018. This translates to states with high SAT participation rates are more likely to have low ACT participation rates, vice versa.

There's an interesting correlation between the score and participation rates. For SAT, the correlation between average total score and participation was -87% in 2017 and -79% in 2018. For ACT, the correlations were -86% in 2017 and -86% in 2018. This shows that the higher the participation, the lower the average total score. This is likely due to students who show greater interest taking the test in cases of low participation and the scores being lower for cases in high participation as the range of student skills become wider.



#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [ ]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels = None, bins=8, sup_title=None,lw=2):
    nrows = int(np.ceil(len(list_of_columns)/2)) # Makes sure you have enough rows
    if len(list_of_columns) > 2:
        figsize=(15,14)
    else:
        figsize=(11,7)
    fig, ax = plt.subplots(nrows=nrows, ncols=2, figsize=figsize) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    fig.suptitle(sup_title, fontsize=20)
    
    mean = dataframe[list_of_columns].mean()
    median = dataframe[list_of_columns].median()
    mode = dataframe[list_of_columns].mode().min()
    
    
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        color_lst = ['green', 'red', 'purple', 'yellow']
        #ax[i].hist(dataframe[column],bins=8, color=color) # feel free to add more settings
        sns.histplot(x=dataframe[column], data=dataframe, bins=bins, color = color_lst[i], ax=ax[i])
        # Set titles, labels, etc here for each subplot
        
        # Axes vline
        #ax[i].axvline(x=0.5, ymin=0.0, ymax=1.0, color='k', linestyle='--', alpha=0.3)
        ax[i].axvline(mean[i], color='magenta', label='mean', lw=lw)
        ax[i].axvline(median[i], color='blue', label='median', lw=lw)
        ax[i].axvline(mode[i], color='black', label='mode', lw=lw)
        
        ax[i].set_title(list_of_titles[i], fontsize=15)
        ax[i].set_xlabel(list_of_xlabels, fontsize=15)
        ax[i].set_ylabel('State Count')
        ax[i].legend(['Mean', 'Median', 'Mode'])

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT
- Math scores for SAT & ACT
- Reading/verbal scores for SAT & ACT

## Plotting the participation rates for SAT 2017 and 2018

In [ ]:
final_sat_act.columns

In [ ]:
subplot_histograms(final_sat_act, ['sat_17_participation', 'sat_18_participation', 'act_17_participation', 'act_18_participation'],
                   ['SAT 2017', 'SAT 2018', 'ACT 2017', 'ACT 2018'], 'Participation rates', 
                   sup_title='SAT and ACT participation rate')

In [ ]:
final_sat_act[['act_17_participation', 'act_18_participation']].agg(['mean', 'median'])

Most of the participation rate is between **0-10%** for **SAT**.

There seems to be about **25%** increase in SAT participation from 2017 to 2018 judging from the *mean values*.

Participation rate for **ACT** is much higher compared to SAT. Also, the participation rates do not differ as much year to year.

## Plotting the math tests for SAT and ACT 2017 and 2018

In [ ]:
subplot_histograms(final_sat_act, ['sat_17_math', 'sat_18_math', 'act_17_math', 'act_18_math'], 
                   ['SAT 17', 'SAT 18', 'ACT 17', 'ACT 18'], 'Math scores', sup_title='SAT and ACT Math scores', bins=15)

Based on the histogram for SAT, the majority's math scores looks to have increased from 525 to 540 and the right end of the curve increase from about 615 to 625
from 2017 to 2018.

For ACT, there's a slight increase in students scoring in Math in 2018 compared to 2017.

## Plotting the reading and verbal scores for SAT 2017 and 2018

In [ ]:
subplot_histograms(final_sat_act, ['sat_17_erw', 'sat_18_erw'],
                   ['SAT 17 reading and writing', 'SAT 18 reading and writing'],
                   'Reading and Writing scores', sup_title='Reading and writing scores', bins=13)

Reading and writing scores have slightly increased from 540 to 550 from 2017 to 2018.

In [ ]:
act_2017.head()

In [ ]:
act_2018.head()

In [ ]:
act_all = pd.merge(act_2017, act_2018, on='state')
act_all.head()

### ACT English scores

In [ ]:
subplot_histograms(act_all, ['act_17_english', 'act_18_english'], 
                   ['ACT 2017', 'ACT 2018'],
                   bins=14,
                   sup_title='ACT English test scores')

In [ ]:
act_all[['act_17_english', 'act_18_english']].agg(['mean','median'])

In [ ]:
act_all[['act_17_english', 'act_18_english']].mode()

Act English scores looks about the same for the years

### ACT Reading scores

In [ ]:
subplot_histograms(act_all, ['act_17_reading', 'act_18_reading'], 
                   ['ACT 2017', 'ACT 2018'],
                   bins=11,
                   sup_title='ACT Reading test scores')

Give that for both histograms, the mean is greater than the median, they're both a right skewed graph as most of the values fall to the left side.

Also the range between 19 ~ 22 becomes more concentrated towards 21 score in 2018 compared to 2017.

### ACT Science scores

In [ ]:
subplot_histograms(act_all, ['act_17_science', 'act_18_science'], 
                   ['ACT 2017', 'ACT 2018'],
                   bins=10,
                   sup_title='ACT Science test scores')

For ACT 2017 science as the mean is greater than the median, it's a right skewed graph as most of the values fall to the left side.

For ACT 2018 science, the histogram becomes more evenly distributed across the range

#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
def lmplots(dataframe, x, y, title, xlabel = None, ylabel = None):    
    sns.lmplot(x=x, y=y, data=dataframe, aspect=2, height=6, ci=None, scatter_kws={'color':'darkred'}, line_kws={'color':'green'})
        # Set titles, labels, etc here for each subplot
    plt.title(title, fontsize=15, pad=15)
#     plt.xticks(ticks=np.arange(450,700,25), labels=[str(x) for x in np.arange(450,700,25)])
    plt.xlabel(xlabel, fontsize=12, labelpad=20)
    plt.ylabel(ylabel, fontsize=12, labelpad=20)

In [ ]:
sat_all.head()

## SAT and ACT 2017 math scores

In [ ]:
lmplots(final_sat_act, 'sat_17_math', 'act_17_math', 
        'SAT and ACT 17 math scores', 'SAT 2017 math scores', 
        'ACT 2017 math scores')

The scatter points are dispersed and a negative line of best fit is plotted. This shows a negative weak correlation between the variables.

## SAT and ACT 2017 verbal and reading scores

In [ ]:
final_sat_act.columns

In [ ]:
lmplots(final_sat_act, 'sat_17_erw', 'act_17_reading', 
             'SAT and ACT 17 verbal and reading scores', 'SAT 2017 reading and writing scores', 
             'ACT 2017 reading and writing scores')

The scatter points are dispersed and a negative line of best fit is plotted. This shows a negative weak correlation between the variables.

## SAT vs. ACT total/composite scores for 2017

In [ ]:
lmplots(final_sat_act, 'sat_17_total', 'act_17_composite', 
             'SAT and ACT 17 total composite scores', 'SAT 2017 total scores', 
             'ACT 2017 composite scores')

The scatter points are dispersed and a negative line of best fit is plotted. This shows a negative weak correlation between the variables.

## Total scores for SAT 2017 vs. 2018

In [ ]:
lmplots(final_sat_act, 'sat_17_total', 'sat_18_total', 
             'SAT 2017 vs SAT 2018 scores', 'SAT 2017 total scores', 
             'SAT 2018 total scores')

Most of the scatter points are clustered together and a positive line of best fit is plotted. This shows a moderately positive correlation between the variables.

## Composite scores for ACT 2017 vs. 2018

In [ ]:
final_sat_act.columns

In [ ]:
lmplots(final_sat_act, 'act_17_composite', 'act_18_composite', 
             'ACT 2017 vs ACT 2018 scores', 'ACT 2017 composite scores', 
             'ACT 2018 composite scores')

Most of the scatter points are clustered together and a positive line of best fit is plotted. This shows a strong positive correlation between the variables.

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

## SAT and ACT participation box plots

In [ ]:
final_sat_act[['sat_17_participation', 'sat_18_participation', 'act_17_participation', 'act_18_participation']].agg(['min', 'max'])

In [ ]:
def boxplots(dataframe, title, xlabel = None, yticklabels = None, palette='coolwarm'):    
    plt.figure(figsize=(11,7))
    sns.boxplot(data=dataframe, orient='h', palette=palette, width=0.5)
        # Set titles, labels, etc here for each subplot
    plt.title(title, fontsize=20, pad=15)
    plt.xlabel(xlabel, labelpad=20, fontsize=15)
    plt.xticks(fontsize=15)
    plt.yticks(ticks=[tick for tick in range(len(dataframe.columns))], labels= yticklabels, fontsize=15)
    plt.show()

In [ ]:
all_part = final_sat_act[['sat_17_participation', 'sat_18_participation', 'act_17_participation', 'act_18_participation']]
boxplots(all_part, 'Participation rates for SAT and ACT test', 'Participation rates in (%)', ['SAT 2017', 'SAT 2018', 'ACT 2017', 'ACT 2018'], palette='viridis')

### Interpreting SAT participation rates

The *participation rate*  for SAT 2018 is **much higher** compared to SAT 2017 given the wider box in SAT 2018. 

The *median participation rate* for SAT 2018 is also **much higher** compared to SAT 2017.

### Interpreting ACT participation rates

In contrast to SAT participation rates, both ACT participation rates in 2017 and 2018 are **much higher compared to SAT**.

There's a slight increase in participation in ACT 2018 compared to ACT 2017.

## SAT evidence based reading and writing box plots

In [ ]:
sat_erw = final_sat_act[['sat_17_erw', 'sat_18_erw']]

In [ ]:
boxplots(sat_erw, 'Evidenced based reading and writing', 'Scores', ['SAT 2017', 'SAT 2018'], palette='viridis')

### Interpreting SAT Evidenced based reading and writing scores

The overall scores for SAT 2018 is slightly lower compared to SAT 2017

## ACT English and Reading box plots

In [ ]:
act_english_reading = final_sat_act[['act_17_english', 'act_18_english', 'act_17_reading', 'act_18_reading']]

In [ ]:
boxplots(act_english_reading,'ACT English and reading scores', 'Scores', 
        ['ACT 17 English', 'ACT 18 English', 'ACT 17 Reading', 'ACT 18 Reading'], palette='magma')

### Interpreting ACT English and Reading scores

**ACT English Scores**

The range has shifted slightly to the right for ACT 2018 and the box plot becoming more right skewed compared to ACT 2017
which means it's becoming less variable.

**ACT Reading Scores**

There is very little difference in the box plots for both years except for the slight decrease in median scores in 2018.

## SAT Math box plots

In [ ]:
final_sat_act[['sat_17_math', 'sat_18_math', 'act_17_math', 'act_18_math']].agg(['min', 'max'])

In [ ]:
sat_math = final_sat_act[['sat_17_math', 'sat_18_math']]

In [ ]:
boxplots(sat_math,'SAT Math scores', 'Scores', ['SAT 2017', 'SAT 2018'], palette='Set2')

### Interpreting SAT Math scores

The lower quartile's range has decreased, the median value has decreased slightly.

The upper quartile has also decreased which makes the score less variable in 2018 compared to 2017.

## ACT Math box plots

In [ ]:
act_math = final_sat_act[['act_17_math', 'act_18_math']]

In [ ]:
boxplots(act_math,'ACT Math scores', 'Scores', ['ACT 2017', 'ACT 2018'], palette='Set1')

### Interpreting ACT Math scores

A positively skewed boxplot with a slight decrease in the change in the middle quartile score.

## ACT Science scores box plots

In [ ]:
act_science = final_sat_act[['act_17_science', 'act_18_science']]

In [ ]:
boxplots(act_science,'SAT Science Scores', 'Scores', ['ACT 2017', 'ACT 2018'], palette='viridis')

### Interpreting ACT Science scores

A positively skewed boxplot with a slight decrease in the change in the middle quartile score.

The lower quartile and the upper quartile score has slightly decreased in 2018 compared to 2017.

The lower whisker is also longer in 2018 compared to 2017 showing that the variability has increased. 

## SAT total score box plots

In [ ]:
sat_total = final_sat_act[['sat_17_total', 'sat_18_total']]

In [ ]:
boxplots(sat_total,'SAT total scores', 'Scores', ['SAT 2017', 'SAT 2018'], palette='viridis')

### Interpreting SAT Total scores

A positively skewed boxplot with a slight decrease in the change in the middle quartile score.

The lower whisker in 2018 is shorter making it less variable compared to 2017.

## ACT composite score box plots

In [ ]:
act_total = final_sat_act[['act_17_composite', 'act_18_composite']]

In [ ]:
boxplots(act_total,'ACT composite scores', 'Scores', ['ACT 2017', 'ACT 2018'], palette='viridis')

### Interpreting ACT composite scores

A positively skewed boxplot with a slight increase in the lower quartile score.

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

The skewness of the data, right / left skewed? 

The spread of the data, your standard deviation. (look at your histogram)

Central Tendency - Mean, Median, Mode

## SAT participation

In [ ]:
subplot_histograms(sat_all, ['sat_17_participation', 'sat_18_participation'], 
                   ['SAT 2017', 'SAT 2018'], 'Participation rates', sup_title='SAT participation rate')

As seen in the histogram, the SAT participation variable is a right skewed as most of the values fall towards the left side and the
mean is greater than the median for 2017 participation histogram.

This is similar to SAT 2018's histogram but it's starting to become more left skewed as more values get distributed towards the right.

In [ ]:
final_sat_act[['sat_17_participation', 'sat_18_participation']].agg(['mean','median','std'])

## SAT ERW

In [ ]:
subplot_histograms(sat_erw, ['sat_17_erw', 'sat_18_erw'],
                   ['SAT 17 reading and writing', 'SAT 18 reading and writing'],
                   'Reading and Writing scores', sup_title='Reading and writing scores', bins=13)

Give that for both histograms, the mean is greater than the median, they're both a right skewed graph.

## SAT Math

In [ ]:
subplot_histograms(sat_math, ['sat_17_math', 'sat_18_math'], 
                   ['SAT 2017', 'SAT 2018'],
                   bins=15,
                   sup_title='SAT math scores')

Give that for both histograms, the mean is greater than the median, they're both a right skewed graph.

## SAT Total

In [ ]:
subplot_histograms(sat_all, ['sat_17_total', 'sat_18_total'], 
                   ['SAT 2017', 'SAT 2018'],
                   bins=13,
                   sup_title='SAT total scores')

In [ ]:
sat_all[['sat_17_total', 'sat_18_total']].agg(['mean', 'median', 'std'])

Give that for both histograms, the mean is greater than the median, they're both a right skewed graph.

## ACT participation

In [ ]:
subplot_histograms(act_all, ['act_17_participation', 'act_18_participation'], 
                   ['ACT 2017', 'ACT 2018'],
                   bins=9,
                   sup_title='ACT total scores')

Give that for both histograms, the mean is less than the median, they're both a left skewed graph as most of the values fall to the right side.

Also the range between 20 ~ 40 is more distributed in 2018 compared to 2017.

## ACT English

In [ ]:
subplot_histograms(act_all, ['act_17_english', 'act_18_english'], 
                   ['ACT 2017', 'ACT 2018'],
                   bins=14,
                   sup_title='ACT English Scores')

Give that for both histograms, the mean is greater than the median, they're both a right skewed graph as most of the values fall to the left side.

Also the range between 23 ~ 26 becomes more concentrated towards the 24 score range in 2018 compared to 2017.

## ACT Math

In [ ]:
subplot_histograms(act_all, ['act_17_math', 'act_18_math'], 
                   ['ACT 2017', 'ACT 2018'],
                   bins=16,
                   sup_title='ACT Math Scores')

In [ ]:
act_all[['act_17_math', 'act_18_math']].agg(['mean', 'median', 'std'])

In [ ]:
act_all[['act_17_math', 'act_18_math']].mode()

Give that for both histograms, the mean is greater than the median, they're both a right skewed graph as most of the values fall to the left side.

Also the range between 23 ~ 24 becomes more distributed across 24 to 25 score in 2018 compared to 2017.

## ACT Reading

In [ ]:
subplot_histograms(act_all, ['act_17_reading', 'act_18_reading'], 
                   ['ACT 2017', 'ACT 2018'],
                   bins=13,
                   sup_title='ACT Reading Scores')

In [ ]:
act_all[['act_17_reading', 'act_18_reading']].agg(['mean', 'median', 'std'])

In [ ]:
act_english_reading.mode()

Give that for both histograms, the mean is greater than the median, they're both a right skewed graph as most of the values fall to the left side.

Also the range between 19 ~ 22 becomes more concentrated towards 21 score in 2018 compared to 2017.

## ACT Science

In [ ]:
subplot_histograms(act_all, ['act_17_science', 'act_18_science'], 
                   ['ACT 2017', 'ACT 2018'],
                   bins=10,
                   sup_title='ACT Science Scores')

For ACT 2017 science as the mean is greater than the median, it's a right skewed graph as most of the values fall to the left side.

For ACT 2018 science, the histogram becomes more evenly distributed.

In [ ]:
act_science.agg(['mean','median','std'])

In [ ]:
act_science.mode()


#### Distributions in the data

In this dataset, each data represents a sample from a population.                        
For example, for ACT math test:
- Population: the test results of all the students who take this test, nation-wide.
- Population mean: is the national average of ACT math test (total scores/total no. of test takers) 
- Sample: the state means of ACT math test. We have 51 samples (51 states)

***According to CLT, we generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?***

Answer: For all the variables in dataframe, they're all sample means of each state and it does not reflect a normally distributed curve as seen in the histograms above.

The histograms does not seem to be centrally distributed around a mean. The participation rates range are between 0-100 and most states are shifted towards having close to 0% participation for SAT and certain states skewed towards 100% participation for ACT. Having distributions away from the mean and towards their upper and lower bounds is not consistent with that of normal.

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer: Under the Central Limit Theorem (CLT), as the score data are state means, such means follow a normal as the sample size increases even if the samples are not normally distributed. However, we must consider sampling bias as students in certain states decide on the whether to participate in a test or not. As seen in the exploratory data analysis, there is high negative correlation between participation rates and scores and the states are shown to have a participation rate between 0% to 100% for each test.

Hence, the average scores for high participation rates should not be compared to average scores for low participation states as this would make the comparison unreliable. To conclude, due to the selection-bias from the students in a state-by-state test participation, the score probability distribution will not lead to a normal.

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

Answer: As the test participation rates do not follow a normal distribution, it would not be correct to make statistical inference between the two variables relying on the normal.

##### Is it appropriate to compare *these* specific SAT and ACT math scores  - can we say students with higher SAT math score is better than those with lower ACT math score, or vice versa?

Why or why not?

In [ ]:
sat_math.describe().loc[:,'sat_18_math']

In [ ]:
act_math.describe().loc[:,'act_18_math']

Answer: No, we cannot. For SAT 2018 math, the range is between 480 to 655 and the average is 556, while for ACT 2018 math, the range is between 18 to 25 and the average is 21. As these two test are conducted under different methodolgies, it would be invalid to compare these tests.

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code:

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

### Difference in participation for SAT for each states

In [ ]:
sat_17_part_state = sat_17_part.sort_values('state')
sat_17_part_state

In [ ]:
sat_18_part_state = sat_18_part.sort_values('state')
sat_18_part_state

In [ ]:
sat_17_18 = sat_17_part_state.merge(sat_18_part_state,on='state')
sat_17_18.head()

In [ ]:
sat_17_18['Difference in %'] = abs(round((sat_17_18['sat_17_participation'] - sat_17_18['sat_18_participation'])/sat_17_18['sat_17_participation'],2))*100
sat_17_18.sort_values('Difference in %',ascending=False).head()

*Illinois* has a large increase in SAT participation with a **1000%** increase, followed by *Colorado* with a **809%** increase in SAT participation 

### Difference in participation for ACT for each states

In [ ]:
act_part = final_sat_act[['state','act_17_participation', 'act_18_participation']].copy()
act_part.head()

In [ ]:
act_part['Difference in %'] = abs(round((act_part['act_17_participation'] - act_part['act_18_participation'])/act_part['act_17_participation'],2))*100
act_part.sort_values('Difference in %', ascending=False).head()

There is a large decline in Colorado and Illinois participation for ACT test from 2017 to 2018. 

In [ ]:
final_sat_act[final_sat_act['state'] == 'Arkansas' ][['state','sat_17_participation', 'sat_17_total', 'sat_18_participation', 'sat_18_total', 
                                                      'act_17_participation', 'act_17_composite', 'act_18_participation', 'act_18_composite']]

Arkansas participation for ACT remains 100% year-to-year and their composite score remains the same while their SAT is close to 0%.

In [ ]:
df_ark_col_lll = final_sat_act[(final_sat_act['state'] == 'Colorado') | (final_sat_act['state'] == 'Illinois') | (final_sat_act['state'] == 'Arkansas')].copy()
df_ark_col_lll.head()

In [ ]:
df_ark_col_lll[['state','sat_17_total', 'sat_18_total', 'act_17_composite', 'act_18_composite']]

Based on my analysis, the three states that I've chosen to investigate Colorado, Illinois and Arkansas. Colorado's SAT participation rate has increased from 11% to 100% from 2017 to 2018 while their ACT participation rate has decreased from 100% to 30%. A similiar trend was found in Illinois where the participation rate has increased from 9% in 2017 to to 99% in 2018 while their ACT participation rate has decreased from 93% to 43%. Arkansas' participation rate for SAT remains the lowest year-to-year while their participation rate for ACT tests is one of the highest.

Colorado state officials switched from the mandatory ACT to the mandatory SAT which resulted in a decrease in ACT participation and an increase in SAT participation. The reason behind the switch is that the Colorado Department of Education (CDE) remarked the exam is more closely aligned to the state's academic standards and to prepare them for life after high school. (1) (2)

According to the Chicago Tribune, Illinois State Board of Education (ISBE) made the switch as well from ACT the SAT after weighing the attributes of both exams and decided SAT's standards was better aligned with Illinois. (3)

For both states, the average scores for SAT decreases significantly as a greater number of students took the test while the average scores for the ACT increased as a result of a smaller group of students taking it. For instance, Colorado's SAT scores went from 1201 in 2017 to 1025 in 2018 while the ACT average score went from 20.8 in 2017 to 23.9 in 2018.

In Arkansas, the ACT is used as an admission requirement to enter colleges or universities which is why there's a a very high participation rate for ACT and it's considered an "ACT state". (4) (5)

### References

* [Testive](https://www.testive.com/colorado-sat-change-2017/) (1) 

* [Chalkbeat](https://co.chalkbeat.org/2017/8/17/21100853/sat-scores-show-mixed-results-on-whether-colorado-juniors-are-on-track-for-college) (2)

* [Chicagotribune](https://www.chicagotribune.com/news/ct-illinois-chooses-sat-met-20160211-story.html) (3)

* [University of Arkansas](https://admissions.uark.edu/apply/test-scores/index.php) (4)

* [Arkansas Department of Education](http://dese.ade.arkansas.gov/divisions/learning-services/assessment/the-act) (5)

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?

After reviewing the state college entrance exam testing policy and examining the datasets, there is clear evidence that the respective state's policy has a huge influence in the student's participation in these tests. Currently, there are 25 states that require students to take the [SAT or ACT](https://www.edweek.org/ew/section/multimedia/states-require-students-take-sat-or-act.html).

For the tests participation to increase, states can choose to enact a policy of making one of the two tests a mandatory high school graduation requirement if they have not done so. This provides benefits to the state's education systems and to the students. One advantage would be that the standardized test can be used to measure the effectiveness of the high-school educational systems. It can also be used to benchmark students and provides students the guidance to complete one common college application requirement. This will be useful in guiding students to consider their school of choice.  

Let's look at Illinois, where students are mandated to take a standardized test to enter colleges and universities. After Illinois State Board of Education (ISBE) reviewed both tests,  ISBE switched to SAT from ACT. This resulted a large decrease in ACT participation and a large increase in SAT participation from 2017 to 2018. These changes in participation were followed by a decrease in average score for the SAT and an increase in ACT average scores. This shows what happens when a state make a test mandatory - a large drop in the participation rate of the other test.

To resolve this, states need to provide incentives to encourage students to take both tests. Incentives such as giving students time off for test taking, subsidizing the fees and integrating the test into high school curriculum can help to remove the barriers.

Research into other factors such as the median household incomes for each state to see if there's any corrleations into the participation rates would be useful to increase the participation rates.